### PR TITLE
[vcpkg_find_acquire_program] Fix extraction, cleanup

### PR DESCRIPTION
**Pull request by dg0yt**
Amends #34881.

`vcpkg_extract_archive` expects the destination to not exist:
https://github.com/microsoft/vcpkg/blob/ee2d2a100103e0f3613c60655dcf15be7d5157b8/scripts/cmake/vcpkg_extract_archive.cmake#L18-L20
but there was no code to ensure that property, leading to  
https://github.com/microsoft/vcpkg/issues/40902#issuecomment-2344467731

In addition, simplify processing of the optional `tool_subdirectory` variable.

### CODE REVIEW

**Comment by BillyONeal** [comment]
> `vcpkg_extract_archive` expects the destination to not exist:

I think this check failing strongly indicates a bug of some sort and I wouldn't want to paper over it like this. For example, in the original bug report, it isn't clear that completely deleting $downloads/tools/nasm is safe or correct.

It seems like we should have been picking something like $/downloads/tools/nasm-$version or $/downloads/nasm/nasm-$version instead?

**Comment by dg0yt** [comment]
> > `vcpkg_extract_archive` expects the destination to not exist:
> 
> I think this check failing strongly indicates a bug of some sort and I wouldn't want to paper over it like this.

IDK. There are only two uses of the function.

> For example, in the original bug report, it isn't clear that completely deleting $downloads/tools/nasm is safe or correct.

I think that `<downloads>/tools` is expected to be under vcpkg's control, including removal of subdirs.
Otherwise, would it really be a location to write to at all?

> It seems like we should have been picking something like $/downloads/tools/nasm-$version or $/downloads/nasm/nasm-$version instead?

What if extraction fails in the middle? This isn't solved with version-based subdirs. Is the user expected to cleanup manually?

IIRC this type of issue has come up a few times, but I don't remember if it was always about nasm/tool_subdirectory. Should tool_subdirectory become mandatory (or at least derived)?

**Comment by BillyONeal** [comment]
> IIRC this type of issue has come up a few times, but I don't remember if it was always about nasm/tool_subdirectory. Should tool_subdirectory become mandatory (or at least derived)?

Good question

**Comment by dg0yt** [comment]
> > IIRC this type of issue has come up a few times, but I don't remember if it was always about nasm/tool_subdirectory. Should tool_subdirectory become mandatory (or at least derived)?
> 
> Good question

NASM examples:
- https://github.com/microsoft/vcpkg/issues/40902#issuecomment-2344467731
- https://github.com/microsoft/vcpkg/issues/35884#issuecomment-1925391828

Other examples:
- ?

**Comment by BillyONeal** [comment]
> > > IIRC this type of issue has come up a few times, but I don't remember if it was always about nasm/tool_subdirectory. Should tool_subdirectory become mandatory (or at least derived)?
> > 
> > 
> > Good question
> 
> NASM examples:
> 
> * [[poco] Build error vcpkg --triplet x64-windows install poco[netssl] fails #40902 (comment)](https://github.com/microsoft/vcpkg/issues/40902#issuecomment-2344467731)
> * [[libjpeg-turbo] Build error on x64-windows #35884 (comment)](https://github.com/microsoft/vcpkg/issues/35884#issuecomment-1925391828)
> 
> Other examples:
> 
> * ?

Given this it seems like the correct change is indeed that `tool_subdirectory` should always be there so that different versions of `nasm` don't try to stomp on each other like this. (It seems like we break people if we ever change the version of nasm?)

Another thing to watch out for is that the downloads directory is expected to be concurrently modified by different running vcpkgs; it isn't one-copy-at-a-time like the installed directory or similar. The existing rename-into-place is intended to preserve this behavior, and it might be broken by the proposed remove-if-the-directory-exists.

**Comment by BillyONeal** [comment]
To clarify about nasm specifically: the nasm zips have `nasm-2.16.01/nasm.exe` inside and we're expecting the version number to come from inside the zip. Which is broken because we are requiring that `downloads/tools/nasm` does not exist.

(Which I argue means that `vcpkg_find_acquire_program` *not* deleting the directory has pointed out this bug to us so I'm still weakly opposed to removing the directory first)

**Comment by dg0yt** [comment]
What can be done:
 - Fix the nasm problem pattern.
   (Probably: subtract to temporary dir, then move.)
 - Improve the message.
   (Give a hint what to check and do.)

**Comment by dg0yt** [comment]
@BillyONeal The modified test failed with the original code and passed with the new code.
Please review. 
I will remove the x64-windows scope when there is a Go.

**Review by MonicaLiu0311 — DISMISSED** [review]

**Comment by MonicaLiu0311** [comment]
Flag `info:reviewed` for further review.

File: scripts/azure-pipelines/azure-pipelines.yml [hunk 1]
```diff
@@ -12,7 +12,7 @@ parameters:
   - name: tripletPattern
     displayName: 'Enable triplets which contain this substring'
     type: string
-    default: '-'
+    default: 'x64-windows$'
```
Comment (dg0yt) [inline scripts/azure-pipelines/azure-pipelines.yml:15]: I will remove the x64-windows scope when there is a Go.

**Review by dg0yt — COMMENTED** [review]
>  Fix the nasm problem pattern.
>  (Probably: subtract to temporary dir, then move.)

Implemented.

> Improve the message.
> (Give a hint what to check and do.)

Not implemented.

File: scripts/cmake/vcpkg_find_acquire_program.cmake [hunk 2]
```diff
@@ -199,6 +199,15 @@ function(vcpkg_find_acquire_program program)
...
             endif()
+        elseif(tool_subdirectory STREQUAL "")
+            # The effective tool subdir is owned by the extracted paths of the archive.
+            # *** This behavior is provided for convenience and short paths. ***
+            # There must be no overlap between different providers of subdirs.
+            # Otherwise tool_subdirectory must be used in order to separate extracted trees.
+            file(REMOVE_RECURSE "${full_subdirectory}.temp")
+            vcpkg_extract_archive(ARCHIVE "${archive_path}" DESTINATION "${full_subdirectory}.temp")
+            file(COPY "${full_subdirectory}.temp/" DESTINATION "${full_subdirectory}")
+            file(REMOVE_RECURSE "${full_subdirectory}.temp")
```
Comment (BillyONeal) [inline scripts/cmake/vcpkg_find_acquire_program.cmake:210]:
```suggestion
            file(RENAME "${full_subdirectory}.temp/" DESTINATION "${full_subdirectory}")
```

?

**Review by BillyONeal — DISMISSED** [review]

File: scripts/cmake/vcpkg_find_acquire_program.cmake (hunk 2 above)
Comment (BillyONeal) [reply scripts/cmake/vcpkg_find_acquire_program.cmake:210]: Oh it can't do that because there's an existing dir.

**Pull request merged**